chain/chain/src/store.rs

```diff
@@ -1745,111 +1780,250 @@ impl<'a> ChainStoreUpdate<'a> {
             .insert((height, shard_id), chunk_hash);
     }
 
-    pub fn clear_old_data_on_height(&mut self, height: BlockHeight) -> Result<(), Error> {
+    // This method may be completely replaced in the next updates
+    pub fn build_refcount_map(&mut self) -> Result<(), Error> {
+        let head = self.head()?;
+        let tail = self.tail()?;
+        self.blocks_refcount = HashMap::new();
+        // All heights from tail including head.height
+        for height in tail..=head.height {
+            let blocks_current_height = match self.get_all_block_hashes_by_height(height) {
+                Ok(blocks_current_height) => {
+                    blocks_current_height.values().flatten().cloned().collect()
+                }
+                _ => vec![],
+            };
+            for block_hash in blocks_current_height.iter() {
+                if let Ok(prev_hash) =
+                    self.get_block(&block_hash).map(|block| block.header.prev_hash)
+                {
+                    *self.blocks_refcount.entry(prev_hash).or_insert(0) += 1;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn get_block_refcount(&self, block_hash: &CryptoHash) -> u64 {
+        self.blocks_refcount.get(block_hash).unwrap_or(&0).clone()
+    }
+
+    pub fn update_tail(&mut self, height: BlockHeight) {
+        self.tail = Some(height);
+    }
+
+    // Clearing block data of `block_hash`, if on a fork.
+    // Clearing block data of `block_hash.prev`, if on the Canonical Chain.
+    pub fn clear_block_data(
+        &mut self,
+        trie: Arc<Trie>,
+        mut block_hash: CryptoHash,
+        is_fork: bool,
+    ) -> Result<(), Error> {
+        let header = match self.get_block_header(&block_hash) {
+            Ok(header) => header.clone(),
+            Err(_) => {
+                // Block data is already cleared
+                panic!("should never happen");
+            }
+        };
```

Comment:
Instead of the match + panic you can use `.expect`:

```rust
let header = self.get_block_header(&block_hash).expect("block data is not expected to be already cleaned").clone();
```

There is also similar block down this function with `get_block`.

Reply:
++